nni/contrib/compression/pruning/taylor_pruner.py

```diff
@@ -0,0 +1,164 @@
+
+    def _collect_data(self) -> Dict[str, Dict[str, torch.Tensor]]:
+        data = defaultdict(dict)
+        for module_name, hooks in self.hooks.items():
+            for target_name, hook in hooks.items():
+                if len(hook.buffer) > 0:
+                    data[module_name][target_name] = hook.buffer[0] / self.training_steps
+        return data
+
+    def _calculate_metrics(self, data: Dict[str, Dict[str, torch.Tensor]]) -> _METRICS:
+        return norm_metrics(p=1, data=data, target_spaces=self._target_spaces)
+
+    def _generate_sparsity(self, metrics: _METRICS) -> _MASKS:
+        return generate_sparsity(metrics, self._target_spaces)
+
+    def _register_hooks(self, evaluator: Evaluator):
+        def collector(buffer: List, target: torch.Tensor) -> Callable[[torch.Tensor], None]:
+            # a factory function, return a tensor hook function for target
+            assert len(buffer) == 0, 'Buffer pass to taylor pruner collector is not empty.'
+
+            def collect_taylor(grad: torch.Tensor):
+                if len(buffer) == 0:
+                    buffer.append(torch.zeros_like(grad))
+                if self._current_step < self.training_steps:
+                    buffer[0] += (target.detach() * grad.detach()).pow(2)
+            return collect_taylor
+
+        hook_list = []
+        for module_name, ts in self._target_spaces.items():
+            for target_name, target_space in ts.items():
+                if is_active_target(target_space):
+                    # TODO: add input/output
+                    if target_space.type is TargetType.PARAMETER:
+                        assert target_space.target is not None
+                        hook = TensorHook(target_space.target,
+                                          target_name,
+                                          functools.partial(collector, target=target_space.target))
+                        hook_list.append(hook)
+                        self.hooks[module_name][target_name] = hook
+                    else:
+                        raise NotImplementedError()
+        evaluator.register_hooks(hook_list)
+
+    def _register_trigger(self, evaluator: Evaluator):
+        assert self.interval_steps >= self.training_steps or self.interval_steps < 0
+        self._remaining_times = self.total_times
+
+        def optimizer_task():
+            self._current_step += 1
+            if self._current_step == self.training_steps:
+                masks = self.generate_masks()
+                self.update_masks(masks)
+                if isinstance(self._remaining_times, int):
+                    self._remaining_times -= 1
+                debug_msg = f'{self.__class__.__name__} generate masks, remaining times {self._remaining_times}'
+                _logger.debug(debug_msg)
+            if self._current_step == self.interval_steps and \
+                (self._remaining_times == 'unlimited' or self._remaining_times > 0):  # type: ignore
```

Comment:
a little confused,  in my opinion, self.interval_steps == self.training_steps and self._remaining_time will be set to 0,  when will this branch logic be executed?

Reply:
in iterative pruning, `self._remaining_time` will be set `>1` by another pruner.